heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -609,6 +610,63 @@ data SimplImpl ps_in ps_out where
     ExprVar (LLVMPointerType w) -> LLVMFieldPerm w sz ->
     SimplImpl (RNil :> LLVMPointerType w) (RNil :> LLVMPointerType w)
 
+  -- | Split an LLVM field permission that points to a known bitvector:
+  --
+  -- > x:[l]ptr((rw,off) |-> eq(bv1++bv2))
+  -- >   -o [l]x:ptr((rw,off) |-> eq(bv1))
+  -- >      * [l]x:ptr((rw,off+len(bv1)) |-> eq(bv2))
+  --
+  -- Note that the definition of @++@ depends on the current endianness.
+  SImpl_SplitLLVMWordField ::
+    (1 <= w, KnownNat w, 1 <= sz1, KnownNat sz1, 1 <= sz2, KnownNat sz2,
+     1 <= (sz2 - sz1), KnownNat (sz2 - sz1)) =>
+    ExprVar (LLVMPointerType w) ->
+    LLVMFieldPerm w sz2 -> BV.BV sz1 -> BV.BV (sz2 - sz1) -> EndianForm ->
+    SimplImpl (RNil :> LLVMPointerType w)
+    (RNil :> LLVMPointerType w :> LLVMPointerType w)
+
+  -- | Concatenate two LLVM field permissions that point to known bitvectors:
+  --
+  -- > [l]x:ptr((rw,off) |-> eq(bv1)) * [l]x:ptr((rw,off+len(bv1)) |-> eq(bv2))
+  -- > -o x:[l]ptr((rw,off) |-> eq(bv1++bv2))
+  --
+  -- Note that the definition of @++@ depends on the current endianness.
+  SImpl_ConcatLLVMWordFields ::
+    (1 <= w, KnownNat w, 1 <= sz1, KnownNat sz1, 1 <= sz2, KnownNat sz2,
+     1 <= (sz1 + sz2), KnownNat (sz1 + sz2)) =>
+    ExprVar (LLVMPointerType w) ->
+    LLVMFieldPerm w sz1 -> BV.BV sz2 -> EndianForm ->
+    SimplImpl (RNil :> LLVMPointerType w :> LLVMPointerType w)
+    (RNil :> LLVMPointerType w)
+
+  -- | Split an LLVM field permission with @true@ contents:
+  --
+  -- > x:[l]ptr((rw,off,sz2) |-> true)
+  -- >   -o [l]x:ptr((rw,off,sz1) |-> true)
+  -- >      * [l]x:ptr((rw,off+sz1,sz2-sz1) |-> true)
+  --
+  -- Note that the definition of @++@ depends on the current endianness.
+  SImpl_SplitLLVMTrueField ::
+    (1 <= w, KnownNat w, 1 <= sz1, KnownNat sz1, 1 <= sz2, KnownNat sz2,
+     1 <= (sz2 - sz1), KnownNat (sz2 - sz1)) =>
+    ExprVar (LLVMPointerType w) -> LLVMFieldPerm w sz2 -> NatRepr sz1 ->
+    NatRepr (sz2 - sz1) ->
```

Comment:
We don't need these `NatRepr`s since we have `KnownNat` instances ([`knownNat`](https://hackage.haskell.org/package/parameterized-utils-2.1.4.0/docs/Data-Parameterized-NatRepr.html#v:knownNat)).

Reply:
Well, I see what you mean, but these arguments are here to determine the `sz1` and `sz2` type arguments. Otherwise uses of this constructor are ambiguous. They also make it easier to write `simplImplIn` and `simplImplOut`. We could change them to `Proxy` instead of `NatRepr` types, but it doesn't actually hurt anything to leave them as is.